src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs

```diff
@@ -287,6 +291,58 @@ public static CorInfoTypeWithMod getParameterType(IntPtr thisHandle, CORINFO_CLA
             return (CorInfoTypeWithMod)corInfoType;
         }
 
+        public struct TypeDescriptor
+        {
+            public uint FieldCount;
+            public CORINFO_FIELD_STRUCT_** Fields; // array of CORINFO_FIELD_STRUCT_*
+        }
+
+        [UnmanagedCallersOnly]
+        public static TypeDescriptor getTypeDescriptor(IntPtr thisHandle, CORINFO_CLASS_STRUCT_* inputType)
+        {
+            var _this = GetThis(thisHandle);
+
+            if (_this.typeDescriptorDict.TryGetValue((IntPtr)inputType, out var typeDescriptor))
+            {
+                return typeDescriptor;
+            }
+
+            TypeDesc type = _this.HandleToObject(inputType);
+
+            uint fieldCount = 0;
+            foreach (var field in type.GetFields())
+            {
+                if (!field.IsStatic)
+                {
+                    fieldCount++;
+                }
+            }
+
+            //TODO-LLVM: change to NativeMemory.Alloc when upgrade to .net6
+            IntPtr allocedMemory = Marshal.AllocHGlobal((int)(sizeof(CORINFO_FIELD_STRUCT_*) * fieldCount));
+            _allocedMemory.Add(allocedMemory);
```

Comment:
I think it is not necessary to give type descriptors process-level lifetime. They should be just fine with compilation-level lifetime.

Nit: `allocedMemory` -> `fieldsArray`.

Reply:
I free them up when shutting down the jit

https://github.com/dotnet/runtimelab/blob/b495aecf7c51dc85cba6e1ee43a3c4f76d5632fb/src/coreclr/tools/aot/ILCompiler.LLVM/Compiler/LLVMCodegenCompilation.cs#L78-L79

If I make them non static and instance fields of `CorInfoImpl` then I'll have to implement the dispose pattern?

Reply:
I would have expected it to be freed somewhere around `CompileMethodCleanup`.

However, now that I look at it again, the descriptors are dictionary-cached, and it is fine for that to have "static" lifetime. So let's leave them as-is.